File: practice/data_structures/linked_lists/detect_cycle/detect_cycle.cpp
/*
  Detect loop in a linked list 
  List could be empty also
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/

int HasCycle(Node * i_head)
{
    if (!i_head)
    {
        return 0;
    }
    
    Node * slow = i_head;
    Node * fast = i_head->next;
    while (slow && fast && (slow != fast))
    {
        slow = slow->next;
        if (!fast->next)
        {
            return 0;
        }
        fast = fast->next->next;
    }
    if (slow && fast && (slow == fast))
    {
        return 1;
    }
    
    return 0;
}

File: practice/data_structures/linked_lists/detect_cycle/detect_cycle.py
"""
 Check if linked list has cycle
 head could be None as well for empty list
 Node is defined as
 
 class Node(object):
 
   def __init__(self, data=None, next_node=None):
       self.data = data
       self.next = next_node

 return 0 if no cycle else return 1
"""

def HasCycle(i_head):
    if not i_head:
        return 0
    
    slow = i_head
    fast = i_head.next
    while slow and fast and slow != fast:
        slow = slow.next
        if not fast.next:
            return 0
        fast = fast.next.next
    if slow and fast and slow == fast:
        return 1
    return 0

